pkg/ottl/ottlfuncs/func_replace_all_patterns.go

```diff
@@ -44,6 +45,41 @@ func createReplaceAllPatternsFunction[K any](_ ottl.FunctionContext, oArgs ottl.
 	return replaceAllPatterns(args.Target, args.Mode, args.RegexPattern, args.Replacement, args.Function)
 }
 
+func applyReplaceAllFunction[K any](ctx context.Context, tCtx K, compiledPattern *regexp.Regexp, fn ottl.Optional[ottl.FunctionGetter[K]], originalValStr string, replacementVal string) (string, error) {
+	var updatedString string
+	updatedString = originalValStr
+	submatches := compiledPattern.FindAllStringSubmatchIndex(updatedString, -1)
+	for _, submatch := range submatches {
+		fullMatch := originalValStr[submatch[0]:submatch[1]]
+		result := compiledPattern.ExpandString([]byte{}, replacementVal, originalValStr, submatch)
+		fnVal := fn.Get()
+		replaceValGetter := ottl.StandardStringGetter[K]{
+			Getter: func(context.Context, K) (any, error) {
+				return string(result), nil
+			},
+		}
+		replacementExpr, errNew := fnVal.Get(&replaceAllPatternFuncArgs[K]{Input: replaceValGetter})
+		if errNew != nil {
+			return "", errNew
+		}
+		replacementValRaw, errNew := replacementExpr.Eval(ctx, tCtx)
+		if errNew != nil {
+			return "", errNew
+		}
+		replacementValStr, ok := replacementValRaw.(string)
+		if !ok {
+			return "", fmt.Errorf("the replacement value must be a string")
+		}
+		if len(submatch) > 2 {
```

Comment:
Can we have a comment that describes what this checks for? If we have no submatches is there something invalid about the value returned from `ExpandString`?

Reply:
```
		// if there are submatches then use the resolved capture group value
		if len(submatch) > 2 {
			updatedString = strings.ReplaceAll(updatedString, fullMatch, replacementValStr)
		}
		// if there are no submatches and the replacement string isn't a capture group then use the replacement value as is
		if !IsGroupNumString(replacementVal) {
			updatedString = strings.ReplaceAll(updatedString, fullMatch, replacementVal)
		}
```

I've made the update